cni/pkg/install-cni/pkg/install/install.go

```diff
@@ -0,0 +1,263 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package install
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"syscall"
+
+	"github.com/coreos/etcd/pkg/fileutil"
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+
+	"istio.io/istio/cni/pkg/install-cni/pkg/config"
+	"istio.io/istio/cni/pkg/install-cni/pkg/constants"
+	"istio.io/istio/cni/pkg/install-cni/pkg/util"
+	"istio.io/pkg/log"
+)
+
+type configFiles struct {
+	kubeconfigFilepath string
+	cniConfigFilepath  string
+}
+
+// Run starts the installation process with given configuration
+func Run(cfg *config.Config) (err error) {
+	var files configFiles
+
+	ctx, cancel := context.WithCancel(context.Background())
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM)
+	go func(sigChan chan os.Signal, cancel context.CancelFunc) {
+		sig := <-sigChan
+		log.Infof("Exit signal received: %s", sig)
+		cancel()
+	}(sigChan, cancel)
+
+	defer func() {
+		if cleanErr := cleanup(cfg, files); cleanErr != nil {
+			if err != nil {
+				err = errors.Wrap(err, cleanErr.Error())
+			} else {
+				err = cleanErr
+			}
+		}
+	}()
+
+	for {
```

Comment:
Is this a one time job or it is a service that keeps running? trying to understand the loop here.

Reply:
By default it is a service that keeps running, but there is a config field `ExecuteOnce` which makes it execute only once

Reply:
I ended up removing this field and behavior since it isn't exposed to the user anyways. I figured if it really is needed, then this can be added in the future.
I found that this "sleep" behavior that existed in the [original shell script](https://github.com/istio/cni/blob/c2c2a7e0c929528f76504e3c799df2c8a7118025/deployments/kubernetes/install/scripts/install-cni.sh#L305) was directly copied from [Calico's install-cni.sh script](https://github.com/projectcalico/cni-plugin/blob/cfe211b008ee79f0ae7cafe0a1ee6ea424509782/k8s-install/scripts/install-cni.sh#L208). The only use case I found for this was in [Calico's the install-cni.sh unit test](https://github.com/projectcalico/cni-plugin/blob/e6905c5ea6b15ae4c228f0f16de26ce1156afeb1/k8s-install/scripts/install_cni_test.go#L43).